webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
+    descriptor_set_layout_binding(1, DT::Sampler, SSF::ALL, false),
+    // Color2 sampler
+    descriptor_set_layout_binding(2, DT::Sampler, SSF::ALL, false),
+];
+
+pub(super) const COMMON_SET_3: &'static [DescriptorSetLayoutBinding] = &[
+    // Color0
+    descriptor_set_layout_binding(0, DT::SampledImage, SSF::ALL, false),
+    // Color1
+    descriptor_set_layout_binding(1, DT::SampledImage, SSF::ALL, false),
+    // Color2
+    descriptor_set_layout_binding(2, DT::SampledImage, SSF::ALL, false),
+];
+
+#[cfg(not(feature = "push_constants"))]
+pub(super) const COMMON_SET_4: &'static [DescriptorSetLayoutBinding] = &[
+    // Locals
+    descriptor_set_layout_binding(0, DT::UniformBuffer, SSF::VERTEX, false),
+];
+
+pub(super) const CLIP_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding(0, DT::CombinedImageSampler, SSF::ALL, true),
+    // RenderTasks
+    descriptor_set_layout_binding(1, DT::CombinedImageSampler, SSF::VERTEX, true),
+    // GpuCache
+    descriptor_set_layout_binding(2, DT::CombinedImageSampler, SSF::ALL, true),
+    // TransformPalette
+    descriptor_set_layout_binding(3, DT::CombinedImageSampler, SSF::VERTEX, true),
+];
+
+pub(super) const PRIMITIVE_SET_1: &'static [DescriptorSetLayoutBinding] = &[
+    // Dither
+    descriptor_set_layout_binding(0, DT::CombinedImageSampler, SSF::ALL, true),
+    // RenderTasks
+    descriptor_set_layout_binding(1, DT::CombinedImageSampler, SSF::VERTEX, true),
+    // GpuCache
+    descriptor_set_layout_binding(2, DT::CombinedImageSampler, SSF::ALL, true),
+    // TransformPalette
+    descriptor_set_layout_binding(3, DT::CombinedImageSampler, SSF::VERTEX, true),
+    // PrimitiveHeadersF
+    descriptor_set_layout_binding(4, DT::CombinedImageSampler, SSF::VERTEX, true),
+    // PrimitiveHeadersI
+    descriptor_set_layout_binding(5, DT::CombinedImageSampler, SSF::VERTEX, true),
+];
+
+pub(super) const PRIMITIVE_SET_0: &'static [DescriptorSetLayoutBinding] = &[
+    // PrevPassAlpha
+    descriptor_set_layout_binding(0, DT::CombinedImageSampler, SSF::ALL, true),
+    // PrevPassColor
+    descriptor_set_layout_binding(1, DT::CombinedImageSampler, SSF::ALL, true),
+];
+
+pub(super) const EMPTY_SET_0: &'static [DescriptorSetLayoutBinding] = &[];
 
 #[derive(Debug, PartialEq, Eq, Copy, Clone, Hash)]
-pub(super) enum ShaderGroup {
-    Brush,
-    Debug,
-    ClipCache,
+pub(super) enum DescriptorGroup {
+    Default,
+    Clip,
+    Primitive,
+}
+
+impl From<ShaderKind> for DescriptorGroup {
+    fn from(kind: ShaderKind) -> Self {
+        match kind {
+            ShaderKind::Cache(VertexArrayKind::Border) | ShaderKind::Cache(VertexArrayKind::LineDecoration)
+                | ShaderKind::DebugColor | ShaderKind::DebugFont
+                | ShaderKind::VectorStencil | ShaderKind::VectorCover => DescriptorGroup::Default,
+            ShaderKind::ClipCache => DescriptorGroup::Clip,
+            _ => DescriptorGroup::Primitive,
```

Comment:
if possible, let's try being more explicit about what those cases are instead of putting an all-catch "_"